File: src/__tests__/provision-ops.test.ts
import fs from "node:fs";
import path from "node:path";
import { describe, expect, it } from "vitest";
import { fingerprint, generateIdentity, verify } from "../shared/crypto.js";
import {
	EnrollOpSchema,
	type FirstRoot,
	firstRootSigningBytes,
	type ProvisionTenant,
	provisionTenantSigningBytes,
	type RemoveTenant,
	removeTenantSigningBytes,
	rosterRequestSigningBytes,
	type SetProfileName,
	setProfileNameSigningBytes,
	signFirstRoot,
	signProvisionTenant,
	signRemoveTenant,
	signRosterRequest,
	signSetProfileName,
	signTrustPendingRequest,
	trustPendingSigningBytes,
	verifyFirstRoot,
	verifyProvisionTenant,
	verifyRemoveTenant,
	verifySetProfileName,
} from "../shared/enrollment.js";
import { ConsoleOpSchema, ProvisioningSchema } from "../shared/schemas.js";
import { assertCanonicalBytes } from "./_canonical-bytes.js";

////////////////////////////////
//  Friend cross-Domain onboarding signing-bytes vectors
//
//  vectors.json is read by BOTH this suite and ProvisionOpsTest.kt (Kotlin), so the
//  hand-authored Kotlin twin (ProvisionOpsCrypto.kt) cannot drift from this TS source:
//  the canonical bytes / signature either runtime derives differently fails one of the two
//  suites. This suite also guards the fixture against a hand-edit (the recorded bytes +
//  signature must reproduce from the live TS reference). The provision / remove ops are
//  operator-signed, first_root is SELF-signed by the fresh owner key, set_profile_name is
//  owner-signed.

interface SignedVec<T> {
	value: T;
	signingBytes: string;
	signingBytesHex: string;
	signingBytesBase64: string;
	signature: string;
}

const vectors = JSON.parse(
	fs.readFileSync(path.join(__dirname, "../../tests/fixtures/provision-ops/vectors.json"), "utf8"),
) as {
	adminSignPub: string;
	adminSignPriv: string;
	adminFingerprint: string;
	friendOwnerSignPub: string;
	friendOwnerSignPriv: string;
	friendOwnerBoxPub: string;
	friendOwnerFingerprint: string;
	provision: SignedVec<ProvisionTenant>;
	removal: SignedVec<RemoveTenant>;
	firstRoot: SignedVec<FirstRoot>;
	rename: SignedVec<SetProfileName>;
	roster: SignedVec<{ signerSignPub: string; proofAt: number; nonce: string }>;
	trustPending: SignedVec<{ signerSignPub: string; proofAt: number; nonce: string }>;
};

describe("provision_tenant vectors (operator-signed)", () => {
	const { adminSignPub, adminSignPriv } = vectors;

	it("reproduces the canonical PROVISION_TENANT_V1 signing bytes", () => {
		const bytes = provisionTenantSigningBytes(vectors.provision.value, adminSignPub);
		assertCanonicalBytes(bytes, vectors.provision);
	});

	it("embeds the operator fingerprint in the signing bytes", () => {
		// The signing bytes bind fingerprint(adminSignPub), not the raw key, so the verifier
		// can recompute it from the signed wrapper's adminSignPub.
		expect(vectors.adminFingerprint).toBe(fingerprint(adminSignPub));
		expect(vectors.provision.signingBytes).toContain(`\n${vectors.adminFingerprint}\n`);
	});

	it("reproduces the recorded signature and verifies it", () => {
		// Ed25519 is deterministic (RFC 8032), so re-signing with the fixed key reproduces the
		// pinned signature byte-for-byte.
		const signed = signProvisionTenant(vectors.provision.value, adminSignPriv, adminSignPub);
		expect(signed.signature).toBe(vectors.provision.signature);
		expect(verifyProvisionTenant(signed, adminSignPub)).toBe(true);
	});

	it("rejects the provision under a different operator key", () => {
		const forged = {
			...signProvisionTenant(vectors.provision.value, adminSignPriv, adminSignPub),
			adminSignPub: "AAAA",
		};
		expect(verifyProvisionTenant(forged, adminSignPub)).toBe(false);
	});

	it("parses a provision_tenant enroll op and rejects a non-slug domainId", () => {
		const signed = signProvisionTenant(vectors.provision.value, adminSignPriv, adminSignPub);
		expect(EnrollOpSchema.safeParse({ kind: "provision_tenant", provision: signed }).success).toBe(true);
		const badSlug = signProvisionTenant(
			{ ...vectors.provision.value, domainId: "Has Spaces" } as ProvisionTenant,
			adminSignPriv,
			adminSignPub,
		);
		expect(EnrollOpSchema.safeParse({ kind: "provision_tenant", provision: badSlug }).success).toBe(false);
	});
});

describe("remove_tenant vectors (operator-signed)", () => {
	const { adminSignPub, adminSignPriv } = vectors;

	it("reproduces the canonical REMOVE_TENANT_V1 signing bytes", () => {
		const bytes = removeTenantSigningBytes(vectors.removal.value, adminSignPub);
		assertCanonicalBytes(bytes, vectors.removal);
	});

	it("reproduces the recorded signature and verifies it", () => {
		const signed = signRemoveTenant(vectors.removal.value, adminSignPriv, adminSignPub);
		expect(signed.signature).toBe(vectors.removal.signature);
		expect(verifyRemoveTenant(signed, adminSignPub)).toBe(true);
	});

	it("a provision signature never replays as a removal (distinct versioned prefix)", () => {
		// Same domainId across provision and removal, but the version prefix differs, so the
		// provision's signature must not verify over the removal bytes.
		const asRemoval: RemoveTenant = {
			domainId: vectors.provision.value.domainId,
			issuedAt: vectors.provision.value.issuedAt,
			nonce: vectors.provision.value.nonce,
		};
		expect(
			verify(removeTenantSigningBytes(asRemoval, adminSignPub), vectors.provision.signature, adminSignPub),
		).toBe(false);
	});

	it("parses a remove_tenant enroll op", () => {
		const signed = signRemoveTenant(vectors.removal.value, adminSignPriv, adminSignPub);
		expect(EnrollOpSchema.safeParse({ kind: "remove_tenant", removal: signed }).success).toBe(true);
	});
});

describe("first_root vectors (self-signed by the fresh owner key)", () => {
	const { friendOwnerSignPub, friendOwnerSignPriv } = vectors;

	it("reproduces the canonical FIRST_ROOT_V1 signing bytes", () => {
		const bytes = firstRootSigningBytes(vectors.firstRoot.value);
		assertCanonicalBytes(bytes, vectors.firstRoot);
	});

	it("reproduces the recorded self-signature and verifies it against the rooted key", () => {
		const signed = signFirstRoot(vectors.firstRoot.value, friendOwnerSignPriv);
		expect(signed.signature).toBe(vectors.firstRoot.signature);
		// The first-root has no separate ownerSignPub: it verifies against firstRoot.ownerSignPub.
		expect(verifyFirstRoot(signed)).toBe(true);
		expect(signed.firstRoot.ownerSignPub).toBe(friendOwnerSignPub);
	});

	it("rejects a first_root whose ownerSignPub was substituted (self-signature breaks)", () => {
		const attacker = generateIdentity();
		const signed = signFirstRoot(vectors.firstRoot.value, friendOwnerSignPriv);
		// Swapping the key inside firstRoot changes the preimage AND the verify key, so the
		// pinned self-signature no longer checks: a captured first-root cannot be re-pointed at
		// an attacker's owner key.
		const forged = { ...signed, firstRoot: { ...signed.firstRoot, ownerSignPub: attacker.sign.pub } };
		expect(verifyFirstRoot(forged)).toBe(false);
	});

	it("rejects a tampered nonce (the one-time QR token is covered by the signature)", () => {
		const signed = signFirstRoot(vectors.firstRoot.value, friendOwnerSignPriv);
		const tampered = { ...signed, firstRoot: { ...signed.firstRoot, nonce: "b3RoZXItbm9uY2U=" } };
		expect(verifyFirstRoot(tampered)).toBe(false);
	});

	it("parses as a console op (a defensive reject), never an enroll op", () => {
		const signed = signFirstRoot(vectors.firstRoot.value, friendOwnerSignPriv);
		// first_root stays a ConsoleOp variant the gateway can defensively reject; the live
		// first-root POSTs DIRECTLY to evie (a pending Domain has no gateway), and it is NOT on
		// the evie enroll surface either (pre-root, the friend has no admission to authenticate).
		expect(ConsoleOpSchema.safeParse({ kind: "first_root", firstRoot: signed }).success).toBe(true);
		expect(EnrollOpSchema.safeParse({ kind: "first_root", firstRoot: signed }).success).toBe(false);
	});
});

describe("set_profile_name vectors (owner-signed)", () => {
	const { friendOwnerSignPub, friendOwnerSignPriv } = vectors;

	it("reproduces the canonical SET_PROFILE_NAME_V1 signing bytes", () => {
		const bytes = setProfileNameSigningBytes(vectors.rename.value, friendOwnerSignPub);
		assertCanonicalBytes(bytes, vectors.rename);
	});

	it("reproduces the recorded signature and verifies it", () => {
		const signed = signSetProfileName(vectors.rename.value, friendOwnerSignPriv, friendOwnerSignPub);
		expect(signed.signature).toBe(vectors.rename.signature);
		expect(verifySetProfileName(signed, friendOwnerSignPub)).toBe(true);
	});

	it("rejects the rename under a different owner key", () => {
		const attacker = generateIdentity();
		const signed = signSetProfileName(vectors.rename.value, friendOwnerSignPriv, friendOwnerSignPub);
		expect(verifySetProfileName(signed, attacker.sign.pub)).toBe(false);
	});

	it("parses a set_profile_name enroll op", () => {
		const signed = signSetProfileName(vectors.rename.value, friendOwnerSignPriv, friendOwnerSignPub);
		expect(EnrollOpSchema.safeParse({ kind: "set_profile_name", rename: signed }).success).toBe(true);
	});
});

describe("ProvisioningSchema pendingTenant (the pending-Domain discriminator)", () => {
	const base = {
		apiUrl: "https://k8s.example:6443",
		caPem: "-----BEGIN CERTIFICATE-----\nMII...\n-----END CERTIFICATE-----",
		saToken: "sa-token-value",
	};

	it("round-trips a blob carrying pendingTenant", () => {
		const blob = {
			...base,
			pendingTenant: { domainId: "guest-network-7", nonce: "bm9uY2UtdmFsdWU=" },
		};
		const parsed = ProvisioningSchema.safeParse(blob);
		expect(parsed.success).toBe(true);
		expect(parsed.success && parsed.data.pendingTenant).toEqual({
			domainId: "guest-network-7",
			nonce: "bm9uY2UtdmFsdWU=",
		});
	});

	it("stays optional: a rooted-Domain blob omits pendingTenant", () => {
		const parsed = ProvisioningSchema.safeParse(base);
		expect(parsed.success).toBe(true);
		expect(parsed.success && parsed.data.pendingTenant).toBeUndefined();
	});

	it("rejects a non-slug domainId and a non-base64 nonce inside pendingTenant", () => {
		expect(
			ProvisioningSchema.safeParse({ ...base, pendingTenant: { domainId: "Bad Domain", nonce: "bm9uY2U=" } })
				.success,
		).toBe(false);
		expect(
			ProvisioningSchema.safeParse({ ...base, pendingTenant: { domainId: "ok-domain", nonce: "not base64!" } })
				.success,
		).toBe(false);
	});
});

describe("roster request proof vectors (console-signed)", () => {
	it("reproduces the canonical ROSTER_V1 bytes + signature (cross-runtime pin)", () => {
		const { signerSignPub, proofAt, nonce } = vectors.roster.value;
		const bytes = rosterRequestSigningBytes(signerSignPub, proofAt, nonce);
		assertCanonicalBytes(bytes, vectors.roster);
		// Deterministic Ed25519: re-signing with the fixed key reproduces the pinned signature.
		expect(signRosterRequest(signerSignPub, proofAt, nonce, vectors.friendOwnerSignPriv)).toBe(
			vectors.roster.signature,
		);
		expect(verify(bytes, vectors.roster.signature, signerSignPub)).toBe(true);
	});
});

describe("trust-pending proof vectors (owner-signed, FLOW-2)", () => {
	it("reproduces the canonical TRUST_PENDING_V1 bytes + signature (cross-runtime pin)", () => {
		const { signerSignPub, proofAt, nonce } = vectors.trustPending.value;
		const bytes = trustPendingSigningBytes(signerSignPub, proofAt, nonce);
		assertCanonicalBytes(bytes, vectors.trustPending);
		expect(signTrustPendingRequest(signerSignPub, proofAt, nonce, vectors.friendOwnerSignPriv)).toBe(
			vectors.trustPending.signature,
		);
		expect(verify(bytes, vectors.trustPending.signature, signerSignPub)).toBe(true);
	});

	it("the distinct version tag stops a roster proof from verifying as a trust-pending query", () => {
		// Same key + proofAt + nonce, but ROSTER_V1 vs TRUST_PENDING_V1, so neither proof crosses over.
		const { signerSignPub, proofAt, nonce } = vectors.trustPending.value;
		const rosterSig = signRosterRequest(signerSignPub, proofAt, nonce, vectors.friendOwnerSignPriv);
		expect(verify(trustPendingSigningBytes(signerSignPub, proofAt, nonce), rosterSig, signerSignPub)).toBe(false);
	});
});

File: android/app/src/test/java/com/atelier_nyaarium/switchboard/crypto/ProvisionOpsTest.kt
package com.atelier_nyaarium.switchboard.crypto

import com.atelier_nyaarium.switchboard.proto.FirstRoot
import com.atelier_nyaarium.switchboard.proto.ProvisionTenant
import com.atelier_nyaarium.switchboard.proto.RemoveTenant
import com.atelier_nyaarium.switchboard.proto.SetDisplayName
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pins the Kotlin friend-onboarding signing (provision_tenant / remove_tenant / first_root /
 * set_display_name) against switchboard's node:crypto, through the same vectors the vitest
 * suite reads (tests/fixtures/provision-ops/vectors.json). If the canonical encoding or the
 * signature scheme diverges, the node-signed artifacts below stop verifying here - which
 * would silently break friend cross-Domain onboarding. Mirrors XDomainLinkTest.
 */
class ProvisionOpsTest {
	private val json = Json { ignoreUnknownKeys = true }

	private fun vectors() =
		json.parseToJsonElement(
			javaClass.classLoader!!.getResourceAsStream("provision-ops/vectors.json")!!.bufferedReader().readText(),
		).jsonObject

	private fun provision(o: JsonObject) =
		ProvisionTenant(
			domainId = o["domainId"]!!.jsonPrimitive.content,
			displayName = o["displayName"]!!.jsonPrimitive.content,
			issuedAt = o["issuedAt"]!!.jsonPrimitive.content.toLong(),
			nonce = o["nonce"]!!.jsonPrimitive.content,
		)

	private fun removal(o: JsonObject) =
		RemoveTenant(
			domainId = o["domainId"]!!.jsonPrimitive.content,
			issuedAt = o["issuedAt"]!!.jsonPrimitive.content.toLong(),
			nonce = o["nonce"]!!.jsonPrimitive.content,
		)

	private fun firstRoot(o: JsonObject) =
		FirstRoot(
			domainId = o["domainId"]!!.jsonPrimitive.content,
			ownerSignPub = o["ownerSignPub"]!!.jsonPrimitive.content,
			ownerBoxPub = o["ownerBoxPub"]!!.jsonPrimitive.content,
			nonce = o["nonce"]!!.jsonPrimitive.content,
			issuedAt = o["issuedAt"]!!.jsonPrimitive.content.toLong(),
		)

	private fun rename(o: JsonObject) =
		SetDisplayName(
			domainId = o["domainId"]!!.jsonPrimitive.content,
			displayName = o["displayName"]!!.jsonPrimitive.content,
			issuedAt = o["issuedAt"]!!.jsonPrimitive.content.toLong(),
			nonce = o["nonce"]!!.jsonPrimitive.content,
		)

	@Test
	fun provisionCanonicalBytesMatchNode() {
		val v = vectors()
		val adminSignPub = v["adminSignPub"]!!.jsonPrimitive.content
		val vec = v["provision"]!!.jsonObject
		val bytes = ProvisionOpsCrypto.provisionSigningBytes(provision(vec["value"]!!.jsonObject), adminSignPub)
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
	}

	@Test
	fun removeCanonicalBytesMatchNode() {
		val v = vectors()
		val adminSignPub = v["adminSignPub"]!!.jsonPrimitive.content
		val vec = v["removal"]!!.jsonObject
		val bytes = ProvisionOpsCrypto.removeSigningBytes(removal(vec["value"]!!.jsonObject), adminSignPub)
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
	}

	@Test
	fun firstRootCanonicalBytesMatchNode() {
		val v = vectors()
		val vec = v["firstRoot"]!!.jsonObject
		val bytes = ProvisionOpsCrypto.firstRootSigningBytes(firstRoot(vec["value"]!!.jsonObject))
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
	}

	@Test
	fun renameCanonicalBytesMatchNode() {
		val v = vectors()
		val ownerSignPub = v["friendOwnerSignPub"]!!.jsonPrimitive.content
		val vec = v["rename"]!!.jsonObject
		val bytes = ProvisionOpsCrypto.setDisplayNameSigningBytes(rename(vec["value"]!!.jsonObject), ownerSignPub)
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
	}

	@Test
	fun rosterRequestCanonicalBytesMatchNode() {
		val v = vectors()
		val vec = v["roster"]!!.jsonObject
		val value = vec["value"]!!.jsonObject
		val signer = value["signerSignPub"]!!.jsonPrimitive.content
		val bytes = ProvisionOpsCrypto.rosterRequestSigningBytes(
			signer,
			value["proofAt"]!!.jsonPrimitive.content.toLong(),
			value["nonce"]!!.jsonPrimitive.content,
		)
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
		// The node-signed proof verifies under the signer key (the twin reproduces it).
		assertTrue(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, signer))
	}

	@Test
	fun trustPendingCanonicalBytesMatchNode() {
		val v = vectors()
		val vec = v["trustPending"]!!.jsonObject
		val value = vec["value"]!!.jsonObject
		val signer = value["signerSignPub"]!!.jsonPrimitive.content
		val bytes = ProvisionOpsCrypto.trustPendingSigningBytes(
			signer,
			value["proofAt"]!!.jsonPrimitive.content.toLong(),
			value["nonce"]!!.jsonPrimitive.content,
		)
		CanonicalBytes.assertCanonicalBytes(bytes, vec)
		assertTrue(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, signer))
		// The distinct version tag means the ROSTER_V1 proof never verifies as a trust-pending query.
		val rosterSig = ProvisionOpsCrypto.signRosterRequest(
			signer,
			value["proofAt"]!!.jsonPrimitive.content.toLong(),
			value["nonce"]!!.jsonPrimitive.content,
			v["friendOwnerSignPriv"]!!.jsonPrimitive.content,
		)
		assertFalse(Crypto.verify(bytes, rosterSig, signer))
	}

	@Test
	fun verifiesNodeSignedAdminOps() {
		val v = vectors()
		val adminSignPub = v["adminSignPub"]!!.jsonPrimitive.content

		val pVec = v["provision"]!!.jsonObject
		val pBytes = ProvisionOpsCrypto.provisionSigningBytes(provision(pVec["value"]!!.jsonObject), adminSignPub)
		assertTrue(Crypto.verify(pBytes, pVec["signature"]!!.jsonPrimitive.content, adminSignPub))

		val rVec = v["removal"]!!.jsonObject
		val rBytes = ProvisionOpsCrypto.removeSigningBytes(removal(rVec["value"]!!.jsonObject), adminSignPub)
		assertTrue(Crypto.verify(rBytes, rVec["signature"]!!.jsonPrimitive.content, adminSignPub))

		// A different admin key must not verify either signature.
		val other = Crypto.generateIdentity()
		assertFalse(Crypto.verify(pBytes, pVec["signature"]!!.jsonPrimitive.content, other.sign.pub))
		assertFalse(Crypto.verify(rBytes, rVec["signature"]!!.jsonPrimitive.content, other.sign.pub))
	}

	@Test
	fun verifiesNodeSignedFirstRoot() {
		val v = vectors()
		val ownerSignPub = v["friendOwnerSignPub"]!!.jsonPrimitive.content
		val vec = v["firstRoot"]!!.jsonObject
		val bytes = ProvisionOpsCrypto.firstRootSigningBytes(firstRoot(vec["value"]!!.jsonObject))
		// The self-signature verifies against the owner key the Domain roots at.
		assertTrue(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, ownerSignPub))
		// Any other key must not verify the node-signed first-root.
		assertFalse(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, Crypto.generateIdentity().sign.pub))
	}

	@Test
	fun verifiesNodeSignedRename() {
		val v = vectors()
		val ownerSignPub = v["friendOwnerSignPub"]!!.jsonPrimitive.content
		val vec = v["rename"]!!.jsonObject
		val bytes = ProvisionOpsCrypto.setDisplayNameSigningBytes(rename(vec["value"]!!.jsonObject), ownerSignPub)
		assertTrue(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, ownerSignPub))
		assertFalse(Crypto.verify(bytes, vec["signature"]!!.jsonPrimitive.content, Crypto.generateIdentity().sign.pub))
	}

	@Test
	fun signsAndVerifiesLocally() {
		val adminOwner = Crypto.generateIdentity()
		val attacker = Crypto.generateIdentity()

		val p = ProvisionTenant("home", "Home Lab", 5000L, "cA==")
		val signedP = ProvisionOpsCrypto.signProvision(p, adminOwner.sign.priv, adminOwner.sign.pub)
		assertTrue(ProvisionOpsCrypto.verifyProvision(signedP, adminOwner.sign.pub))
		assertFalse(ProvisionOpsCrypto.verifyProvision(signedP, attacker.sign.pub))
		// A tampered displayName must not verify.
		assertFalse(ProvisionOpsCrypto.verifyProvision(signedP.copy(provision = p.copy(displayName = "Evil")), adminOwner.sign.pub))

		val r = RemoveTenant("home", 6000L, "cg==")
		val signedR = ProvisionOpsCrypto.signRemove(r, adminOwner.sign.priv, adminOwner.sign.pub)
		assertTrue(ProvisionOpsCrypto.verifyRemove(signedR, adminOwner.sign.pub))
		assertFalse(ProvisionOpsCrypto.verifyRemove(signedR, attacker.sign.pub))

		// The distinct prefixes mean a provision signature does not verify as a removal over the
		// same fields (and the reverse), so neither can be replayed as the other.
		val crossRemoval = RemoveTenant(p.domainId, p.issuedAt, p.nonce)
		assertFalse(Crypto.verify(ProvisionOpsCrypto.removeSigningBytes(crossRemoval, adminOwner.sign.pub), signedP.signature, adminOwner.sign.pub))
	}

	@Test
	fun firstRootSignsAndVerifiesLocally() {
		val owner = Crypto.generateIdentity()
		val attacker = Crypto.generateIdentity()
		val f = FirstRoot("home", owner.sign.pub, owner.box.pub, "bm9uY2U=", 7000L)
		val signed = ProvisionOpsCrypto.signFirstRoot(f, owner.sign.priv)
		assertTrue(ProvisionOpsCrypto.verifyFirstRoot(signed))
		// Re-pointing the rooted ownerSignPub breaks the self-signature.
		assertFalse(ProvisionOpsCrypto.verifyFirstRoot(signed.copy(firstRoot = f.copy(ownerSignPub = attacker.sign.pub))))
		// A tampered nonce (the one-time QR token) breaks the self-signature.
		assertFalse(ProvisionOpsCrypto.verifyFirstRoot(signed.copy(firstRoot = f.copy(nonce = "b3RoZXI="))))
	}

	@Test
	fun renameSignsAndVerifiesLocally() {
		val owner = Crypto.generateIdentity()
		val attacker = Crypto.generateIdentity()
		val r = SetDisplayName("home", "My Network", 8000L, "bg==")
		val signed = ProvisionOpsCrypto.signSetDisplayName(r, owner.sign.priv, owner.sign.pub)
		assertTrue(ProvisionOpsCrypto.verifySetDisplayName(signed, owner.sign.pub))
		assertFalse(ProvisionOpsCrypto.verifySetDisplayName(signed, attacker.sign.pub))
		assertFalse(ProvisionOpsCrypto.verifySetDisplayName(signed.copy(rename = r.copy(displayName = "Hijacked")), owner.sign.pub))
	}
}
